File: quicksort/quicksort.js



function quickSort(arr, low=0, high=arr.length - 1) {
    if (low < high) {
        pIdx = partition(arr, low, high)
        quickSort(arr, low, pIdx - 1)
        quickSort(arr, pIdx + 1, high)
    }

    return arr
}


function partition(arr, low, high){
    let i = low - 1
    pivot = high

    for (let j = low; j <= high - 1; j++) {
        if (arr[j] <= arr[pivot]) {
            i++
            [arr[i], arr[j]] = [arr[j], arr[i]]
        }
    }
    i++
    [arr[pivot], arr[i]] = [arr[i], arr[pivot]]

    return i
}

let newArr = [10, 16, 8, 12, 15, 6, 3, 9, 5]
let arr = [33, 15, 10]

console.log(quickSort(newArr, 0, newArr.length - 1))

File: quicksort/quicksort.py
import random
from types import new_class

#arr = []

#for i in range(1000):
#    arr.append(random.randint(0, i))

#print(arr)

def quicksort(arr):

    if len(arr) < 2:
        return arr

    else:
        pivot = arr[0]
        less = [i for i in arr[1:] if i <= pivot]
        
        greater = [i for i in arr[1:] if i > pivot]

        return quicksort(less) + [pivot] + quicksort(greater)


#print(quicksort(arr))


def partition(arr, low, high):
    indexOfSmaller = low - 1
    pivot = arr[high]

    for j in range(low, high):

        if arr[j] < pivot:

            indexOfSmaller += 1
            arr[indexOfSmaller], arr[j] = arr[j], arr[indexOfSmaller]
    
    indexOfSmaller += 1
    arr[indexOfSmaller], arr[high] = arr[high], arr[indexOfSmaller]

    return indexOfSmaller


def updatedQuickSort(arr, low, high):
    if len(arr) == 1:
        return arr
    if low < high:
        partitionIndex = partition(arr, low, high)

        updatedQuickSort(arr, low, partitionIndex - 1)
        updatedQuickSort(arr, partitionIndex + 1, high)

newArr = [10, 16, 8, 12, 15, 6, 3, 9, 5]

updatedQuickSort(newArr, 0, len(newArr) - 1)

print(newArr)
